File: packages/ts/src/topology.ts
/** Topology (monorepo/workspace) resolution + package-manager / build-tool inference. */
import { readFileSync } from "node:fs";
import { join } from "node:path";
import type { Topology } from "./types.js";
import type { SignalMatrix, TopologyDef } from "./matrix.js";
import { fileMatches } from "./signals.js";

function safeRead(root: string, rel: string): string | null {
  try {
    return readFileSync(join(root, rel), "utf8");
  } catch {
    return null;
  }
}

function matchTopologyDef(def: TopologyDef, root: string, files: string[]): string | null {
  if (def.glob) {
    const hit = files.find((f) => fileMatches(f, def.glob!));
    return hit ?? null;
  }
  if (def.marker) {
    const { file, contains, jsonKey } = def.marker;
    const hit = files.find((f) => fileMatches(f, file));
    if (!hit) return null;
    const content = safeRead(root, hit);
    if (content == null) return null;
    if (contains && content.includes(contains)) return hit;
    if (jsonKey) {
      try {
        const obj = JSON.parse(content) as Record<string, unknown>;
        if (Object.prototype.hasOwnProperty.call(obj, jsonKey)) return hit;
      } catch {
        return null;
      }
    }
    return null;
  }
  return null;
}

export function resolveTopology(matrix: SignalMatrix, root: string, files: string[]): Topology {
  const matched: { def: TopologyDef; signal: string }[] = [];
  for (const def of matrix.topology) {
    const signal = matchTopologyDef(def, root, files);
    if (signal != null) matched.push({ def, signal });
  }
  if (matched.length === 0) return { type: "single", tool: null, signals: [] };

  const hasMonorepo = matched.some((m) => m.def.type === "monorepo");
  const type = hasMonorepo ? "monorepo" : "workspace";
  // Prefer a tool whose type matches the resolved topology type, else the first match.
  const primary = matched.find((m) => m.def.type === type) ?? matched[0]!;
  const signals = [...new Set(matched.map((m) => m.signal))].sort();
  return { type, tool: primary.def.tool, signals };
}

export function inferPackageManagers(matrix: SignalMatrix, files: string[]): string[] {
  const tools = new Set<string>();
  for (const pm of matrix.packageManagers) {
    if (files.some((f) => fileMatches(f, pm.glob))) tools.add(pm.tool);
  }
  return [...tools].sort();
}

export function inferBuildTools(matrix: SignalMatrix, files: string[]): string[] {
  const tools = new Set<string>();
  for (const bt of matrix.buildTools) {
    if (files.some((f) => fileMatches(f, bt.glob))) tools.add(bt.tool);
  }
  return [...tools].sort();
}

File: packages/py/src/repo_fingerprint/topology.py
"""Topology resolution + package-manager / build-tool inference."""
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from .signals import file_matches
from .types import Topology


def _safe_read(root: str, rel: str) -> str | None:
    try:
        return (Path(root) / rel).read_text(encoding="utf-8")
    except Exception:
        return None


def _match_topology_def(defn: dict[str, Any], root: str, files: list[str]) -> str | None:
    glob = defn.get("glob")
    if glob:
        return next((f for f in files if file_matches(f, glob)), None)
    marker = defn.get("marker")
    if marker:
        hit = next((f for f in files if file_matches(f, marker["file"])), None)
        if not hit:
            return None
        content = _safe_read(root, hit)
        if content is None:
            return None
        if marker.get("contains") and marker["contains"] in content:
            return hit
        if marker.get("jsonKey"):
            try:
                obj = json.loads(content)
            except Exception:
                return None
            if isinstance(obj, dict) and marker["jsonKey"] in obj:
                return hit
    return None


def resolve_topology(matrix: dict[str, Any], root: str, files: list[str]) -> Topology:
    matched: list[tuple[dict[str, Any], str]] = []
    for defn in matrix["topology"]:
        sig = _match_topology_def(defn, root, files)
        if sig is not None:
            matched.append((defn, sig))
    if not matched:
        return Topology(type="single", tool=None, signals=[])

    has_monorepo = any(d["type"] == "monorepo" for d, _ in matched)
    topo_type = "monorepo" if has_monorepo else "workspace"
    primary = next((d for d, _ in matched if d["type"] == topo_type), matched[0][0])
    signals = sorted({s for _, s in matched})
    return Topology(type=topo_type, tool=primary["tool"], signals=signals)


def infer_package_managers(matrix: dict[str, Any], files: list[str]) -> list[str]:
    tools: set[str] = set()
    for pm in matrix["packageManagers"]:
        if any(file_matches(f, pm["glob"]) for f in files):
            tools.add(pm["tool"])
    return sorted(tools)


def infer_build_tools(matrix: dict[str, Any], files: list[str]) -> list[str]:
    tools: set[str] = set()
    for bt in matrix["buildTools"]:
        if any(file_matches(f, bt["glob"]) for f in files):
            tools.add(bt["tool"])
    return sorted(tools)
